aws/rust-runtime/aws-http/src/user_agent.rs

```diff
@@ -79,29 +90,58 @@ impl AwsUserAgent {
             language_metadata: LanguageMetadata {
                 lang: "rust",
                 version: "1.50.0",
-                extras: vec![],
+                extras: Default::default(),
             },
             exec_env_metadata: None,
-            additional_metadata: vec![],
+            feature_metadata: Vec::new(),
+            config_metadata: Vec::new(),
+            framework_metadata: Vec::new(),
+            app_name: None,
         }
     }
 
+    #[doc(hidden)]
+    /// Adds feature metadata to the user agent.
+    pub fn with_feature_metadata(mut self, metadata: FeatureMetadata) -> Self {
+        self.feature_metadata.push(metadata);
+        self
+    }
+
+    #[doc(hidden)]
+    /// Adds config metadata to the user agent.
+    pub fn with_config_metadata(mut self, metadata: ConfigMetadata) -> Self {
+        self.config_metadata.push(metadata);
+        self
+    }
+
+    #[doc(hidden)]
+    /// Adds framework metadata to the user agent.
+    pub fn with_framework_metadata(mut self, metadata: FrameworkMetadata) -> Self {
+        self.framework_metadata.push(metadata);
+        self
+    }
+
+    /// Sets the app name for the user agent.
+    pub fn with_app_name(mut self, app_name: AppName) -> Self {
+        self.app_name = Some(app_name);
+        self
+    }
+
```

Comment:
fine for now, but if we ever actually want to use these, I don't think you can easily get an owned user agent from the property bag so we probably need versions of these that accept `&mut self`

Reply:
Oh, I was under the impression that anyone using these would be using the Smithy client and insert the property upfront. I suppose they could use a custom middleware to modify it though, so this makes sense.